# writing; exceptions
_COSC 101, Introduction to Computing I, 2021-11-10_

## Announcements
* Homework 8 due tomorrow at 5pm
* Exam 3 next week Wednesday
    * Study materials posted on Moodle

## Outline
* Warm-up
* Writing files
* Debugging file reading/writing
* Exceptions

## Warm-up
Write a function called `odd_lines` that takes the name of a file and returns a list containing the 1st, 3rd, 5th, etc. line of the file.


```python
def odd_lines(filename):
    '''
    Construct a list containing the 1st, 3rd, 5th, etc. line of the file
    
    >>> odd_lines('warm-up.txt')
    ['First\\n', 'Third\\n', 'Fifth\\n']
    '''
    infile = open(filename, 'r')
    line = infile.readline() # Read first line
    lines = []
    while line != '': # Repeat until end of file
        lines = lines + [line]
        line = infile.readline() # Skip a line
        line = infile.readline()
    infile.close()
    return lines
```

## Writing files

### Overview
* What does a program need to do to write data to a file?
    * Open the file in write mode
    * Call the `write` function on the file object
    * Call the `close` function on the file object


```python
myfile = open('myfile.txt', 'w')
myfile.write('Hello, file!\n')
myfile.close()
```

* You must explicitly include newline characters in the string passed to write if you want the next string you write to appear on the next line of the file
    * write **does not** automatically include newlines like print does
* If you do not call close, part of the data may be lost
    * Python stores data in a temporary buffer before writing it to the file; close ensures all data in the buffer is added to the file

## Debugging file reading/writing
Assume the file `months.txt` contains the following data:
```
January
February
March
April
May
June
July
August
September
October
November
December
```

*Identify and correct the error(s) in each of the following programs.*


```python
#1) Print the contents of the file
months = open('months.txt', 'r')
for line in months:
    print(line)
```

    January
    
    February
    
    March
    
    April
    
    May
    
    June
    
    July
    
    August
    
    September
    
    October
    
    November
    
    December
    



```python
#1) Print the contents of the file
# CORRECTED
months = open('months.txt', 'r')
for line in months:
    print(line.strip()) # Need to strip newline to avoid extra blank lines
months.close() # Need to close file
```

    January
    February
    March
    April
    May
    June
    July
    August
    September
    October
    November
    December



```python
#2) Print the number of characters in each month
months = open('months.txt', 'r')
while months.readline() != '':
    line = months.readline()
    print(len(line))
months.close()
```

    9
    6
    5
    7
    8
    9



```python
#2) Print the number of characters in each months
# CORRECTED
months = open('months.txt', 'r')
line = months.readline() # Priming read
while line != '': # Check if previously read line was empty
    print(len(line))
    line = months.readline() # Read next line
months.close()
```

    8
    9
    6
    6
    4
    5
    5
    7
    10
    8
    9
    9



```python
#3) Store the number of lines in `months.txt` in a new file called `count.txt`
months = open('months.txt', 'w')
data = months.read()
months.close()
size = open('size.txt', 'w')
size.write('lines: '+str(len(data)))
size.close()
```


    ---------------------------------------------------------------------------

    UnsupportedOperation                      Traceback (most recent call last)

    /var/folders/bd/35sz70xs5837wszrx11bw1yc0000gn/T/ipykernel_1537/855805968.py in <module>
          1 #3) Store the number of lines in `months.txt` in a new file called `count.txt`
          2 months = open('months.txt', 'w')
    ----> 3 data = months.read()
          4 months.close()
          5 size = open('size.txt', 'w')


    UnsupportedOperation: not readable



```python
#3) Store the number of lines in `months.txt` in a new file called `count.txt`
# CORRECTED
months = open('months.txt', 'r') # Open file for reading
data = months.readlines() # Get a list of lines
months.close()
size = open('size.txt', 'w')
size.write('lines: '+str(len(data)))
size.close()
```

## Exceptions

\#1) *What happens if we run the following program and the user enters `one`?*


```python
num = int(input('Enter a number: '))
```

    Enter a number: one



    ---------------------------------------------------------------------------

    ValueError                                Traceback (most recent call last)

    <ipython-input-32-e5e4786827d2> in <module>
    ----> 1 num = int(input('Enter a number: '))
    

    ValueError: invalid literal for int() with base 10: 'one'


* Exception — a signal that something exceptional (i.e., unexpected or unrecoverable) has happened
* *What causes an exception?* — invalid input, invalid math (e.g., divide by zero), attempt to open a non-existent file, etc.
* *Why do we need exceptions?* — a function may not know what to do when something unexpected occurs, so it needs to notify the calling function
* Exceptions breaks a program’s flow of execution
    * Function where exception occurred stops executing — e.g., the `int` function stops executing
    * Exception is passed to calling function
        * If calling function includes code to “recover” from the exception then execution resumes in the **exception handling** code
        * Otherwise, the exception is passed to the preceding function
        * Again, if preceding function includes code to “recover” from the exception then execution resumes in the exception handling code
        * Otherwise, the exception is passed to the preceding function
        * If the program does not have any code to handle the exception, then it crashes

### Example


```python
def func3():
    num = int(input('Enter a number: '))
    return num

def func2():
    input_num = func3()
    return input_num * 2

def func1():
    doubled_num = func2()
    return doubled_num * 2

def main():
    quadrupled_num = func1()
    print("Quadrupled:", quadrupled_num)

main()
```

* Flow of execution
    * `main` calls `func1` which calls `func2` which calls `func3` which calls `input` and `int`
    * `int` raises an exception
    * `func3` has no code to recover from (i.e., handle) the exception
    * Attempt to resume execution in caller (`func2`)
    * `func2` has no code to handle the exception
    * Attempt to resume execution in preceding caller (`func1`)
    * `func1` has no code to handle the exception
    * Attempt to resume execution in preceding caller (`main`)
    * `main` has no code to handle the exception
    * Attempt to resume execution outside of function
    * No code outside of function to handle the exception -> program crashes
* *How could we prevent our program from crashing if the user enters invalid input?*
    * Validate the user’s input **before** converting to an int and ask for new input if the user did not enter a valid integer
    * **Catch** the exception and ask for new input
* Best practice: do not use exception handling (`try-except`) if you can easily write code to check for and recover from errors
    * For example, validating the user's input before converting to an int is better than catching an exception

## Extra practice
\#1) *What is the contents of `seasons.txt` after the program has finished?*


```python
seasons = [['Winter', 'cold', 'snowy'], 
           ['Spring', 'warm', 'rainy'], 
           ['Summer', 'hot', 'sunny'], 
           ['Autumn', 'cool', 'windy']]
outfile = open('seasons.txt', 'w')
outfile.write('In Hamilton:\n')
for season in seasons:
    outfile.write(season[0] + ' is...')
    outfile.write(' and '.join(season[1:]) + '\n')
outfile.close()
```

```
In Hamilton:
Winter is...cold and snowy
Spring is...warm and rainy
Summer is...hot and sunny
Autumn is...cool and windy
```
